### ISSUE TITLE
Change disk image size

### ISSUE THREAD

**Issue by Valexr** [utmapp/UTM#2636]
Сan i expand the disk space after create image & installing system on it?
Ubuntu not enough space...🤓

**Comment by wxw-matt** [comment]
Instead of expanding the disk (disk file) itself, you can add a new disk (disk file) to the Ubuntu and mount the new disk as a new mount point or a subfolder.
https://www.cyberciti.biz/faq/mount-drive-from-command-line-ubuntu-linux/
If you are using LVM to manage your space, you can follow the [article](https://www.cyberciti.biz/faq/howto-add-disk-to-lvm-volume-on-linux-to-increase-size-of-pool/) to extend the storage

**Comment by Valexr** [comment]
tx @wxw-matt 🤓

**Comment by conath** [comment]
Duplicate of #16

**Comment by johnsaigle** [comment]
Found a simple answer in the support Discord that worked for me:

`qemu-img resize /path/to/your/something.qcow2 +SIZE`

Then use e.g. `gparted` within the VM to expand your drive to use the new space.

**Comment by fffx** [comment]
For UTM on macOS:
```
qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
```

see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image

**Comment by jcrf84** [comment]
> For UTM on macOS:
> 
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> ```
> 
> see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image

hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?

**Comment by f1yankees** [comment]
> > For UTM on macOS:
> > ```
> > qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image
> 
> hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?

Same, this breaks the image.  For me, it creates a "display output is not active" error at login

**Comment by samuela** [comment]
> For UTM on macOS:
> 
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> ```
> 
> see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image

Thank you! This is great. Furthermore, I found https://codenotary.com/blog/enlarge-a-disk-and-partition-of-any-linux-vm-without-a-reboot/ useful to resize my linux guest partitions and file system without having to shutdown.

**Comment by xJkit** [comment]
> > For UTM on macOS:
> > ```
> > qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> > ```  
> >     
> > see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image
> 
> hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?

@jcrf84 @f1yankees  


It works on my M1 MacBook Air but with a subtle approach.

I have no qcow2 named `disk-0.qcow2`, but you can find your own by opening the package contents. 

###  Find your qcow2 file.
What you should to is go to the path `~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm` ,right click on your UTM file and choose `Show Package Contents`. Afterwards, find the qcow2 inside the `Images/<YOUR OWN NAME>.qcow2`, copy that name and do the follwoing command again.

### Increase your disk volume 

```
qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/<YOUR OWN NAME>.qcow2 +100G 
```

where +100G is your option.

### Allocate the disk
Go to the VM(WIndows 11 ARM), then go `Settings > Create and format hard disk partitions`, you will see the extra spaces you just created.

**Comment by therene** [comment]
> > > For UTM on macOS:
> > > ```
> > > qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image
> > 
> > 
> > hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?
> 
> @jcrf84 @f1yankees
> 
> It works on my M1 MacBook Air but with a subtle approach.
> 
> I have no qcow2 named `disk-0.qcow2`, but you can find your own by opening the package contents.
> 
> ### Find your qcow2 file.
> What you should to is go to the path `~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm` ,right click on your UTM file and choose `Show Package Contents`. Afterwards, find the qcow2 inside the `Images/<YOUR OWN NAME>.qcow2`, copy that name and do the following command again.
> 
> ### Increase your disk volume
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/<YOUR OWN NAME>.qcow2 +100G 
> ```
> 
> where +100G is your option.
> 
> ### Allocate the disk
> Inside the VM(WIndows 11 ARM), go to `Settings > Create and format hard disk partitions`, and you will see extra spaces you just created.

ok, but after that, when I got ,,image resized" in terminal, - what I need to do? Cause in vm windows 11 I see my extra space in ,,create and format hard disk partition", I can do the partition of new disk space, but in the and, when I was try to install big file (60gb game), system inform that ,,im run out of space and choose other disk". Im choosing new disk but nothings change .. :/

**Comment by robigan** [comment]
Out of curiosity, where are you guys getting the proper qemu-img binary from? I installed utm from homebrew but no qemu-* (or more specifically, no qemu-img) binaries are found on my path??

**Comment by robigan** [comment]
Interesting, I thought that adding binaries to the path is against apple ToS when using apps on the app store

**Comment by niranjanakella** [comment]
I found the solution to this problem.

- First, you need to install the qemu-utils using the homebrew `brew install qemu`
- Then you need to execute the following command to increase your image size
`qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/<Your_Image_Name>.qcow2 +10G ` 
- Where '+10G' is the size to which you would like to increase the image to.
- If you would like to refer to it directly from the finder, don't forget to toggle the hidden folders & files option using "Command+Shift+."

**Comment by verisence** [comment]
> > > For UTM on macOS:
> > > ```
> > > qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image
> > 
> > 
> > hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?
> 
> @jcrf84 @f1yankees
> 
> It works on my M1 MacBook Air but with a subtle approach.
> 
> I have no qcow2 named `disk-0.qcow2`, but you can find your own by opening the package contents.
> 
> ### Find your qcow2 file.
> What you should to is go to the path `~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm` ,right click on your UTM file and choose `Show Package Contents`. Afterwards, find the qcow2 inside the `Images/<YOUR OWN NAME>.qcow2`, copy that name and do the following command again.
> 
> ### Increase your disk volume
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/<YOUR OWN NAME>.qcow2 +100G 
> ```
> 
> where +100G is your option.
> 
> ### Allocate the disk
> Inside the VM(WIndows 11 ARM), go to `Settings > Create and format hard disk partitions`, and you will see extra spaces you just created.

I had to restart the Windows 11 ARM VM for the changes to take efffect.

**Comment by francois-spectre** [comment]
> > > For UTM on macOS:
> > > ```
> > > qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/disk-0.qcow2 +10G 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > see also: https://superuser.com/questions/24838/is-it-possible-to-resize-a-qemu-disk-image
> > 
> > 
> > hey, had tried it several times on my M1 MacBook Air with an Windows 11 VM but somehow it won't work, could you help me?
> 
> @jcrf84 @f1yankees
> 
> It works on my M1 MacBook Air but with a subtle approach.
> 
> I have no qcow2 named `disk-0.qcow2`, but you can find your own by opening the package contents.
> 
> ### Find your qcow2 file.
> What you should to is go to the path `~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm` ,right click on your UTM file and choose `Show Package Contents`. Afterwards, find the qcow2 inside the `Images/<YOUR OWN NAME>.qcow2`, copy that name and do the following command again.
> 
> ### Increase your disk volume
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<YOU VM>.utm/Images/<YOUR OWN NAME>.qcow2 +100G 
> ```
> 
> where +100G is your option.
> 
> ### Allocate the disk
> Inside the VM(WIndows 11 ARM), go to `Settings > Create and format hard disk partitions`, and you will see extra spaces you just created.

This worked for me except it was for ubuntu and I had to follow this tutorial to resize the partition:
https://packetpushers.net/ubuntu-extend-your-default-lvm-space/#:~:text=Select%20your%20%2Fdev%2Fsda3%20partition,to%20confirm%2C%20and%20hit%20ENTER.

**Comment by git0uttaHereIci** [comment]
`qemu-img` resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/Star\ Wars\ Yoda.utm/Images/data.qcow2 +10G`
`Image resized.`

**Comment by git0uttaHereIci** [comment]
> `qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/Star\ Wars\ Yoda.utm/Images/data.qcow2 +10G` `Image resized.`

I did this but am having trouble resizing the image with parted. I don't have enough diskspace to install gparted or grow part.

**Comment by mattpopovich** [comment]
For anyone looking to increase their disk image size with an Ubuntu VM, this is 100% the article that you've been looking for. Step by step instructions for a M1 Mac using UTM: https://linux.afnom.net/install_apple/m1.html

I've been trying to figure this out for the last couple hours, and that article got me the solution in ~5mins. 

Hope this helps others as much as it helped me.

**Comment by mig8447** [comment]
For me the macOS command above didn't work, I instead did

```
qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<VM_NAME>.utm/Data/disk0.img +10G
```

**Comment by duongel** [comment]
> For me the macOS command above didn't work, I instead did
> 
> ```
> qemu-img resize ~/Library/Containers/com.utmapp.UTM/Data/Documents/<VM_NAME>.utm/Data/disk0.img +10G
> ```

Worked for me, but the VM does not recognise the increased disk space. Any ideas?

**Comment by larsut** [comment]
> 

Updated link: https://linux.afnom.net/install_apple/apple_silicon.html#expanding-your-utm-vms-storage-a-nameutm-storage-expanda

**Comment by philmarion** [comment]
(...2 years laters ) yes, you can!
1) if your UTM apps is from github, there's no qemu-img command, but : 
     `brew install qemu`   will provide it . then,  `qemu-img /path/disk.qcow2 +10G` will resize +10Go 
2) with qemu and an iso of [gparted](https://gparted.org/):
```
qemu-system-x86_64 -m 1024 \ 
 -boot d -cdrom /download_path/gparted-live-1.5.0-6-amd64.iso \
 -usb \
 -device usb-tablet \ 
 -drive file=/path2/disk.qcow2,if=virtio
```
(path2=/Users/me/Library/Containers/com.utmapp.UTM/Data/Documents/myVM.utm/Data)
note: on linux, I use to resize fs in cli with virt-resize (from libguestfs-tools), but it seems there's no way to get it on macos ?

**Comment by petapika** [comment]
I tried a lot of the above with an Arch Linux install as an IMG disk using Apple Virtualization and I was unable to resize my disk. 

I was finally able to do it using CloneZilla operating on an imported IMG--> GCOW from Arch and cloning to a fresh disk created in UTM. I moved the new GCOW disk to a new VM without Apple Virtualization and booted from it with a new size and everything worked great.

**Comment by msart-it-vision** [comment]
I tried to Clone the HD but any App i tried to install on my Win11 ARM in UTM failed either to install or run.

So i tried the mentioned CloneZilla Arm Version as ISO and worked from the first Boot as it should.

Addet NVME Bigger Space in UTM and USB CD with the CloneZilla ISO moved the CD Up in the List booted direct and choosed Pro (r Option to Scale the Source to Destination Size was allready selected.

Without the ability to use the bigger space and scale the Partition it makes no sense, so cloning with a bootable iso is working best.

xD

**Comment by msart-it-vision** [comment]
You can just go to the Edit of the VM in UTM:
![image](https://github.com/utmapp/UTM/assets/89912235/c97393d0-8030-47cf-99dc-bd5a661311d8)
Then Click on the HD in the Sidebar and Resize via the GUI:
![image](https://github.com/utmapp/UTM/assets/89912235/57b8ca95-0062-46a6-81af-5a621a0a391c)
![image](https://github.com/utmapp/UTM/assets/89912235/eeb037a3-902d-426f-a381-7f09cca1936c)
![image](https://github.com/utmapp/UTM/assets/89912235/b4eeaab8-9294-49b7-8fdc-cdb93c30d79b)
![image](https://github.com/utmapp/UTM/assets/89912235/0779eb9b-6c77-44e5-9261-051a26557541)

**Comment by sontt9** [comment]
1. Resize on the UTM GUI
2. Going to the VM, then
```shell
# Determine which partition to resize
fdisk -l

# Go into the parted command
parted /dev/vda
# then print the list of partition
print

# resize the partition, with 1 refier to partition number, 100% tell parted to expand to fill the remaining space
resizepart 2 100%

# Exit parted
quit

# Resize the filesystem
resize2fs /dev/vda2

# Verify the new size
df -h
``` 

Worked with Macbook Pro M1, and Arch VM

**Comment by lloeki** [comment]
> Is there a way to do that with Apple virtualization?

@5alv1 I opened a dedicated issue for that not-QEMU case: https://github.com/utmapp/UTM/issues/6891

You'll find how to do it manually over there.

**Comment by midhunkrishna** [comment]
Might not be related to the root cause of the thread, but this is what comes up with a google search for "utm disk resize not working / issues". 

Note: This is written and summarized using Claude. 




# macOS APFS boot-disk resize — troubleshooting write-up

**Date:** 2026-06-30 · **Machine:** Apple Silicon, macOS 15.6.1, internal 137.4 GB (128 GiB) SSD

## TL;DR

A 128 GiB SSD was showing only ~63 GB total with ~9–11 GB free. Three things were wrong, in layers:

1. The **APFS boot container was created at only ~63 GB** — ~68 GB of the SSD was **unallocated**.
2. A **stray standalone Recovery partition** sat between the boot container and that free space, blocking any grow.
3. Underneath, the **GPT partition-map "usable range" was stale**, capped at the old ~64 GiB boundary — so even after removing the recovery partition, the container refused to grow into the physically-free space.

**What fixed it:** remove the stray Recovery partition (required disabling SIP from Recovery OS) → `sudo diskutil repairDisk disk0` (rewrote the stale partition-map geometry) → `sudo diskutil apfs resizeContainer disk2 0` (grew to full size).

**Result:** `/` went from ~63 GB / ~9 GB free → **137 GB / ~86 GB free.**

---

## Symptom

`df` reported a ~63 GB volume with ~9 GB free, but the machine has a 128 GB SSD. `du` suggested ~84 GB used — the numbers didn't reconcile.

## Diagnosis

`diskutil list` on the physical disk revealed the real layout:

```
/dev/disk0  (internal, physical)   137.4 GB        ← the "128 GB" SSD (128 GiB)
  disk0s1  Apple_APFS_ISC            0.5 GB
  disk0s2  Apple_APFS (boot)        62.8 GB        ← macOS + Data + everything lived here
  disk0s3  Apple_APFS_Recovery       5.4 GB        ← stray standalone recovery
  (free space)                      68.7 GB        ← unallocated, invisible to df
```

So everything was crammed into a 62.8 GB container; the rest of the SSD was unpartitioned. The real macOS recovery already existed *inside* the boot container (`disk3s3 (Recovery)`), so the standalone `disk0s3` recovery was a redundant leftover (likely from a clone/restore/migration).

---

## Step-by-step: what was tried, and what happened

| # | Action | Result | Verdict |
|---|---|---|---|
| 1 | `sudo diskutil apfs resizeContainer disk3 0` | **Error -69519** "a gap is required…/grow beyond another partition" | ✗ — the recovery partition was in the way |
| 2 | `sudo diskutil apfs deleteContainer disk2` (delete recovery) | "not allowed… APFS Recovery Container… override with `csrutil disable`" | ✗ — SIP protects recovery containers |
| 3 | Boot **Recovery OS** → `csrutil disable` → delete the recovery partition → reboot | Recovery partition removed; **74.1 GB free**; **device IDs renumbered** (boot container became `disk2`) | ✔ |
| 4 | `sudo diskutil apfs resizeContainer disk2 0` | Grew **only +5.4 GB** (62.8 → 68.2 GB) and stopped at the old recovery boundary (~64 GiB) | ◑ partial — clue that a hard boundary existed at 64 GiB |
| 5 | `diskutil apfs resizeContainer disk2 limits` | Reported **Maximum = 136.9 GB** | ⚠ misleading — the true max was reachable in theory but not in practice |
| 6 | **Reboot**, then `sudo diskutil apfs resizeContainer disk2 0` | **Error -69743** "new size must be different than existing" (it thought max = current) | ✗ |
| 7 | `sudo diskutil apfs resizeContainer disk0s2 136g` | "size must not be greater than the partition map allocatable size" | ✗ — **`g` = GiB (binary)**, so `136g` = 146 GB > max |
| 8 | `sudo diskutil apfs resizeContainer disk0s2 136800000000` (explicit bytes) | Computed max 136.9 GB, aligned, then **Error -69519** "beyond the end of partition map usable space" | ✗ |
| 9 | `sudo diskutil apfs resizeContainer disk0s2 130000000000` (leave 7 GB tail) | Same **-69519** | ✗ — ruled out the "gap too small" theory |
| 10 | `sudo diskutil verifyDisk disk0` | "The partition map appears to be OK" | ⚠ consistency-only check — did **not** catch the geometry problem |
| 11 | `sudo gpt show /dev/disk0` | Clean-looking map; the grow stopped at **block 134,217,688 ≈ 2²⁷ = 64 GiB** | 🔍 root-cause clue: stale GPT *usable range* from an original ~64 GB setup |
| 12 | `sudo dd if=/dev/disk0 … skip=1` (read GPT header) | Empty output (`od: cannot skip past end of input`) | ✗ — buffered whole-disk device not seekable while mounted |
| 13 | `sudo dd if=/dev/rdisk0 …` (raw device) | **`Operation not permitted`** | ✗ — macOS blocks raw boot-disk access (Apple Silicon + SIP/Disk-Arbitration). Killed the `gdisk`/raw-edit plan |
| 14 | `sudo diskutil repairDisk disk0` | Repaired the **partition map** (rewrote the stale usable-range / secondary GPT) | ✔ — **the actual fix** |
| 15 | `sudo diskutil apfs resizeContainer disk2 0` | Container grew to **136.9 GB** | ✔ 🎯 |

**Final:** `df -H /` → `137G total, 11G used, 86G free`.

---

## What worked

```bash
# 1. (in Recovery OS) disable SIP, then delete the stray recovery partition
csrutil disable
diskutil apfs deleteContainer <recovery-container>     # done from Recovery
# (reboot back to macOS; re-enable SIP later from Recovery if desired)

# 2. repair the partition-map geometry  ← the key step
sudo diskutil repairDisk disk0

# 3. grow the boot container to fill the disk
sudo diskutil apfs resizeContainer disk2 0
df -H /
```

## What did NOT work (and why)

- **`resizeContainer … 0` (auto-max)** — mis-computed the target as the current size after partition edits (errors -69743 / partial grow). Don't trust `0` after a partition change.
- **Explicit sizes (`136800000000`, `130000000000`)** — failed with **-69519** because the underlying GPT *usable range* was stale; no target size could get past the bad geometry.
- **`gdisk` / raw `dd` edits** — impossible: macOS denies raw boot-disk access (`/dev/rdisk0` → *Operation not permitted*). Only Apple's `diskutil` / Disk Utility can modify the boot disk's map.
- **`verifyDisk`** — passed, so it gave false reassurance; it checks primary/backup *consistency*, not whether the usable range matches the physical disk.
- **Disk Utility GUI / create-temp-partition trick** — not needed in the end; `repairDisk` was the cleaner, direct fix (though the GUI would likely also have rewritten the map).

## Key lessons / gotchas

- **`df` on macOS shows the APFS *container*, not the physical disk.** Unallocated space is invisible there — check `diskutil list <physical>`.
- **Standalone Recovery partitions block container grows** and are **SIP-protected** — deletion needs `csrutil disable` from Recovery OS.
- **Device identifiers (`disk2`/`disk3`…) renumber** after partition changes — re-check `diskutil list` every time; prefer the stable partition id (`disk0s2`).
- **`diskutil` size suffixes are binary (`g` = GiB, not GB).** Use explicit **bytes** to avoid the off-by-7%.
- **`verifyDisk` ≠ `repairDisk`.** For a "the map looks fine but won't grow / -69519" problem, **`repairDisk` is the tool** — it rewrites the GPT geometry (stale usable-range, secondary header) that `verifyDisk` only consistency-checks.
- **You can't hand-edit the boot disk's GPT on a running Apple-Silicon Mac.** Raw access is denied; do partition surgery via `diskutil`/Disk Utility, and use **Recovery OS** for anything the running system refuses.
- **Error code cheat-sheet:** `-69519` = partition-map gap / beyond usable space; `-69743` = requested size equals current.

## Final state

```
/dev/disk2s1s1   137G   11G   86G   12%   /
APFS Container disk2:  136.9 GB ceiling, ~85.5 GB free
disk0s2 (Apple_APFS):  136.9 GB  — fills the disk, no trailing free space
```
